src/Text/Pandoc/Readers/JATS.hs

```diff
@@ -321,24 +352,28 @@ parseBlock (Elem e) =
                                      (TableFoot nullAttr [])
          isEntry x  = named "entry" x || named "td" x || named "th" x
          parseElement = filterChildren isEntry
-         sect n = do isbook <- gets jatsBook
-                     let n' = if isbook || n == 0 then n + 1 else n
-                     labelText <- case filterChild (named "label") e of
-                                    Just t -> (<> ("." <> space)) <$>
-                                              getInlines t
-                                    Nothing -> return mempty
-                     headerText <- case filterChild (named "title") e `mplus`
-                                        (filterChild (named "info") e >>=
-                                            filterChild (named "title")) of
-                                      Just t  -> (labelText <>) <$>
-                                                  getInlines t
-                                      Nothing -> return mempty
-                     oldN <- gets jatsSectionLevel
-                     modify $ \st -> st{ jatsSectionLevel = n }
-                     b <- getBlocks e
-                     let ident = attrValue "id" e
-                     modify $ \st -> st{ jatsSectionLevel = oldN }
-                     return $ headerWith (ident,[],[]) n' headerText <> b
+         parseWithHeader n mBlocks = do 
+                            isbook <- gets jatsBook
+                            let n' = if isbook || n == 0 then n + 1 else n
+                            labelText <- case filterChild (named "label") e of
+                                            Just t -> (<> ("." <> space)) <$>
+                                                      getInlines t
+                                            Nothing -> return mempty
+                            headerText <- case filterChild (named "title") e `mplus`
+                                                (filterChild (named "info") e >>=
+                                                    filterChild (named "title")) of
+                                              Just t  -> (labelText <>) <$>
+                                                          getInlines t
+                                              Nothing -> return labelText
+                            oldN <- gets jatsSectionLevel
+                            modify $ \st -> st{ jatsSectionLevel = n }
+                            blocks <- mBlocks
```

Comment:
Ah @kamoe , I'm relatively new to haskell so apologies if my understanding/vocabulary is a bit off: 

My understanding is that the monadic action wouldn't run until we invoke it using the `<-`, so by passing `blocks` wrapped in a monad 
```
parseWithHeader (sectionLevel+1) (getBlocks e)
```
`(getBlocks e)` wouldn't actually evaluate until execution reach L:370, which is basically how this function worked before. I feel relatively confident this is correct because the jats-reader test has a bunch of nested headers that evaluate correctly